vm.c

```diff
@@ -4443,6 +4448,7 @@ Init_vm_objects(void)
     vm->loading_table = st_init_strtable();
     vm->ci_table = st_init_table(&vm_ci_hashtype);
     vm->frozen_strings = st_init_table_with_size(&rb_fstring_hash_type, 10000);
+    vm->cc_refinement_table = st_init_table(&vm_cc_refinement_hashtype);
```

Comment:
```suggestion
    vm->cc_refinement_table = st_init_numtable();
```

I think you can just use a standard `numtable`, and get rid of your `hash` and `cmp` functions.

Reply:
Ah, I see! I force-pushed the fix.

Reply:
Does numtable compute hash values from address?
If so, it might be necessary to rebuild and recalculate the hash values after GC compaction.
Alternatively, using a unique object_id as in the original implementation could result in stable hash values.

Which is better: rebuilding or the original approach?
If rebuilding, where should it happen? I’ll investigate tomorrow.

However, this is my first time writing C code, so I might be saying something incorrect.

Reply:
> Does numtable compute hash values from address?

Yes it does.

> If so, it might be necessary to rebuild and recalculate the hash values after GC compaction.

Indeed, I thought it being declared as a weak table would be enough, but looking closer at the code, it update the value in place, ignoring the hash.

So we indeed either need an address independent hash, or to use a more complex way to update that table after compaction.

> Alternatively, using a unique object_id as in the original implementation could result in stable hash values.

`object_id` insert that object in two other tables (`obj_to_id` and `id_to_obj`), so it's not ideal.

Reply:
> or to use a more complex way to update that table after compaction.

One simple way could just be to rebuild it entirely after compaction. Iterate on the old one and insert in a new one, then swap them.

Reply:
You should probably reuse a `compare_by_identity` Hash. Search for hidden_identity_hash_new() usages. Some other refinement logic already uses it.

Reply:
I updated the code to perform rehashing (deletion and reinsertion) within the foreach loop, following the approach used in `vm_weak_table_gen_ivar_foreach`.

Also, `value` was added for compatibility with vm_weak_table_foreach_update_weak_key, but it's no longer needed after switching to foreach. So now I'm passing `1` instead of `cc`, as the value is unused.
Hope #13074 gets merged soon 👍

Reply:
I considered the copy-then-swap approach, but if no objects move, it wastes memory by duplicating data unnecessarily.
I'm going with a naive implementation for now, though I'm not confident it's the right choice.
If GC compaction moves most cc, I'd prefer copy-then-swap.
I'd appreciate any insights on how GC compaction behaves.

Reply:
copy-then-swap seems better after all. It's fast when there are few entries, and if there are many, ST_REPLACE will likely happen anyway.
@byroot  Updated the code, would appreciate a review 😄

Reply:
Thanks to everyone who's already reviewed this-no issues have been spotted so far. @peterzhu2118 , when you have a moment, could you take a look at the GC-related changes as well?